File: frontend/src/services/authService.ts
import api from './api';
import axios from 'axios';
import { ApiResponse, LoginRequest, LoginResponse, RegisterRequest, User } from '../types';

const getApiErrorMessage = (error: unknown, fallback: string): string => {
  if (axios.isAxiosError(error)) {
    const responseData = error.response?.data as any;
    const detailMessage = responseData?.error?.details?.[0]?.message;
    return detailMessage || responseData?.error?.message || error.message || fallback;
  }
  return error instanceof Error ? error.message : fallback;
};

export const authService = {
  async login(credentials: LoginRequest): Promise<LoginResponse> {
    try {
      const response = await api.post<ApiResponse<LoginResponse>>('/auth/login', credentials);
      if (response.data.success && response.data.data) {
        localStorage.setItem('token', response.data.data.token);
        return response.data.data;
      }
      throw new Error(response.data.error?.message || 'Login failed');
    } catch (error) {
      throw new Error(getApiErrorMessage(error, 'Login failed'));
    }
  },

  async getCurrentUser(): Promise<User> {
    const response = await api.get<ApiResponse<{ user: User }>>('/auth/me');
    if (response.data.success && response.data.data) {
      return response.data.data.user;
    }
    throw new Error('Failed to get current user');
  },

  async sendVerificationCode(email: string): Promise<{ message: string; verification_challenge_id?: string }> {
    const normalizedEmail = String(email || '').trim();
    if (!normalizedEmail) {
      throw new Error('请输入有效邮箱地址');
    }

    try {
      const response = await api.post<ApiResponse<{ message: string; verification_challenge_id?: string }>>('/auth/send-verification-code', {
        email: normalizedEmail,
      });
      if (response.data.success && response.data.data) {
        return response.data.data;
      }
      throw new Error(response.data.error?.message || 'Failed to send verification code');
    } catch (error) {
      throw new Error(getApiErrorMessage(error, 'Failed to send verification code'));
    }
  },

  async register(payload: RegisterRequest): Promise<LoginResponse> {
    try {
      const response = await api.post<ApiResponse<LoginResponse>>('/auth/register', payload);
      if (response.data.success && response.data.data) {
        localStorage.setItem('token', response.data.data.token);
        return response.data.data;
      }
      throw new Error(response.data.error?.message || 'Registration failed');
    } catch (error) {
      throw new Error(getApiErrorMessage(error, 'Registration failed'));
    }
  },

  logout() {
    localStorage.removeItem('token');
  },

  isAuthenticated(): boolean {
    return !!localStorage.getItem('token');
  },

  async sendPhoneCode(phone: string): Promise<{ message: string; verification_challenge_id?: string }> {
    try {
      const response = await api.post<ApiResponse<{ message: string; verification_challenge_id?: string }>>('/auth/send-phone-code', { phone });
      return response.data.data || { message: '验证码已发送' };
    } catch (error) {
      throw new Error(getApiErrorMessage(error, '发送验证码失败'));
    }
  },

  async bindPhone(phone: string, code: string): Promise<void> {
    try {
      const response = await api.post<ApiResponse<{ message: string }>>('/auth/bind-phone', { phone, code });
      if (!response.data.success) throw new Error(response.data.error?.message || '绑定失败');
    } catch (error) {
      throw new Error(getApiErrorMessage(error, '绑定失败'));
    }
  },

  // 账号注销（密码确认；成功即返回，调用方负责登出）
  async deleteAccount(password: string): Promise<ApiResponse<{ message: string }>> {
    const response = await api.post<ApiResponse<{ message: string }>>('/auth/delete-account', { password });
    return response.data;
  },
};

File: client-desktop/src/HoYoMusic.Desktop.Infrastructure/Services/AuthService.cs
using System.Net.Http.Json;
using System.Net.Http.Headers;
using System.Text.Json;
using HoYoMusic.Desktop.Core.Abstractions;
using HoYoMusic.Desktop.Core.Contracts;
using HoYoMusic.Desktop.Core.Models;

namespace HoYoMusic.Desktop.Infrastructure.Services;

public sealed class AuthService : IAuthService
{
    private readonly HttpClient _httpClient;
    private readonly ITokenStore _tokenStore;
    private static readonly JsonSerializerOptions JsonOptions = new(JsonSerializerDefaults.Web);

    public AuthService(HttpClient httpClient, ITokenStore tokenStore)
    {
        _httpClient = httpClient;
        _tokenStore = tokenStore;
    }

    public async Task<AuthSession> LoginAsync(LoginRequest request, CancellationToken cancellationToken = default)
    {
        using var response = await _httpClient.PostAsJsonAsync("auth/login", new
        {
            identifier = request.Identifier,
            password = request.Password,
        }, JsonOptions, cancellationToken);

        var envelope = await response.Content.ReadFromJsonAsync<ApiEnvelope<LoginResponseData>>(JsonOptions, cancellationToken);
        if (envelope is null)
        {
            throw new ApiException("Failed to parse login response.");
        }

        if (!response.IsSuccessStatusCode || !envelope.Success || envelope.Data is null || string.IsNullOrWhiteSpace(envelope.Data.Token))
        {
            var message = envelope.Error?.Message ?? "Login failed.";
            throw new ApiException(message, envelope.Error?.Code);
        }

        await _tokenStore.SaveTokenAsync(envelope.Data.Token, cancellationToken);

        return new AuthSession
        {
            Token = envelope.Data.Token,
            User = envelope.Data.User,
        };
    }

    public async Task<string?> SendRegistrationVerificationCodeAsync(SendVerificationCodeRequest request, CancellationToken cancellationToken = default)
    {
        using var response = await _httpClient.PostAsJsonAsync("auth/send-verification-code", new
        {
            email = request.Email.Trim(),
        }, JsonOptions, cancellationToken);

        var envelope = await response.Content.ReadFromJsonAsync<ApiEnvelope<SendVerificationCodeResponseData>>(JsonOptions, cancellationToken);
        if (!response.IsSuccessStatusCode || envelope?.Success != true)
        {
            throw new ApiException(envelope?.Error?.Message ?? "Failed to send verification code.", envelope?.Error?.Code);
        }

        return envelope.Data?.VerificationChallengeId;
    }

    public async Task<AuthSession> RegisterAsync(RegisterRequest request, CancellationToken cancellationToken = default)
    {
        using var response = await _httpClient.PostAsJsonAsync("auth/register", new
        {
            username = request.Username.Trim(),
            email = request.Email.Trim(),
            verification_challenge_id = request.VerificationChallengeId.Trim(),
            verification_code = request.VerificationCode.Trim(),
            password = request.Password,
            confirm_password = request.ConfirmPassword,
        }, JsonOptions, cancellationToken);

        var envelope = await response.Content.ReadFromJsonAsync<ApiEnvelope<LoginResponseData>>(JsonOptions, cancellationToken);
        if (envelope is null)
        {
            throw new ApiException("Failed to parse register response.");
        }

        if (!response.IsSuccessStatusCode || !envelope.Success || envelope.Data is null || string.IsNullOrWhiteSpace(envelope.Data.Token))
        {
            throw new ApiException(envelope.Error?.Message ?? "Register failed.", envelope.Error?.Code);
        }

        await _tokenStore.SaveTokenAsync(envelope.Data.Token, cancellationToken);
        return new AuthSession
        {
            Token = envelope.Data.Token,
            User = envelope.Data.User,
        };
    }

    public Task<string?> GetSavedTokenAsync(CancellationToken cancellationToken = default)
        => _tokenStore.GetTokenAsync(cancellationToken);

    public async Task<AuthUser?> GetCurrentUserAsync(CancellationToken cancellationToken = default)
    {
        var token = await _tokenStore.GetTokenAsync(cancellationToken);
        if (string.IsNullOrWhiteSpace(token))
        {
            return null;
        }

        using var request = new HttpRequestMessage(HttpMethod.Get, "auth/me");
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);

        using var response = await _httpClient.SendAsync(request, cancellationToken);
        if (response.StatusCode == System.Net.HttpStatusCode.Unauthorized)
        {
            await _tokenStore.ClearTokenAsync(cancellationToken);
            throw new ApiException("登录状态已过期，请重新登录。", "UNAUTHORIZED");
        }

        var envelope = await response.Content.ReadFromJsonAsync<ApiEnvelope<CurrentUserResponseData>>(JsonOptions, cancellationToken);

        if (!response.IsSuccessStatusCode || envelope?.Success != true)
        {
            var message = envelope?.Error?.Message ?? "Failed to load current user.";
            throw new ApiException(message, envelope?.Error?.Code);
        }

        return envelope.Data?.User;
    }

    public async Task ChangePasswordAsync(ChangePasswordRequest request, CancellationToken cancellationToken = default)
    {
        var token = await _tokenStore.GetTokenAsync(cancellationToken);
        if (string.IsNullOrWhiteSpace(token))
        {
            throw new ApiException("Not authenticated.", "MISSING_TOKEN");
        }

        using var message = new HttpRequestMessage(HttpMethod.Post, "auth/change-password")
        {
            Content = JsonContent.Create(new
            {
                currentPassword = request.CurrentPassword,
                newPassword = request.NewPassword,
            }),
        };
        message.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);

        using var response = await _httpClient.SendAsync(message, cancellationToken);
        var envelope = await response.Content.ReadFromJsonAsync<ApiEnvelope<object>>(JsonOptions, cancellationToken);
        if (!response.IsSuccessStatusCode || envelope?.Success != true)
        {
            if (response.StatusCode == System.Net.HttpStatusCode.Unauthorized)
            {
                await _tokenStore.ClearTokenAsync(cancellationToken);
                throw new ApiException("Session expired. Please login again.", "UNAUTHORIZED");
            }

            throw new ApiException(envelope?.Error?.Message ?? "Failed to change password.", envelope?.Error?.Code);
        }
    }

    public Task LogoutAsync(CancellationToken cancellationToken = default)
        => _tokenStore.ClearTokenAsync(cancellationToken);
}

